### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by himdel**
Shopping cart for service catalog items

Images:

Add to shopping cart under Service Catalogs > item detail
sc-add-button

Notification that item was added (stays on the same screen)
sc-added

Shopping cart modal when empty
sc-empty

Shopping cart modal with some items (updated, twice)
sc_full

Top right bar shopping cart indicator - empty and full
sc_icon_empty
sc_icon_count

still missing: 

   * [x] API changes - https://github.com/ManageIQ/manageiq/pull/7971
   * [x] RBAC - show shoppign cart icon only if `svc_catalog_provision`
   * order history - separate PR

### CODE REVIEW

File: shopping_cart.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Myself and Dan had a meeting with Tina and GM yesterday abt shopping cart backend work
```
Comment (himdel) [inline shopping_cart.txt]: TODO remove

**Comment by himdel** [comment]
@epwinchell , @serenamarie125 .. could you please look at the images if it's OK?

**Comment by epwinchell** [comment]
@himdel I tend to use "btn-danger" very sparingly. I'd suggest "btn-default" instead for "reset" and "delete". Also, please use "pficon-delete" instead of "fa-trash-o". I also, think we need some styling between cart items.

**Comment by himdel** [comment]
@epwinchell thanks, fixed both btn-danger and the icon.. Any tips on how to style the rows?  I was hoping to use pf-data-list, but it seems like it can't compare items based on identity, so it looked like a pain to implement deletion (or generate ids), but maybe I can rewrite that once I add persistence and actually have some ids.

**Comment by epwinchell** [comment]
@himdel I'll will look around.

**Comment by epwinchell** [comment]
@himdel How about: https://www.patternfly.org/widgets/#list-group ?

**Comment by himdel** [comment]
> list-group 

Works for me :) I'll add it in the next push..

**Comment by serenamarie125** [comment]
I agree with @epwinchell on using list-group as well as using pficon-delete.  Can you tag me when you have that available?  Would love to see what it looks like and possibly suggest something similar for PatternFly.  Thanks!

**Comment by himdel** [comment]
@serenamarie125 , @epwinchell I updated to use list-group (updated the relevant screenshot), so from that point of view it should be ready..

Still some API changes needed before merging - right now, I can't remove a service request, nor update it to set process=true... (looking into it..)

File: client/app/states/marketplace/details/details.state.js [hunk 2]
```diff
@@ -113,11 +113,14 @@
...
+        name: vm.serviceTemplate.name,
+
         dialogUrl: dialogUrl,
         serviceTemplateId: serviceTemplate.id,
+
         data: {
           action: 'order',
-          resource: dialogFieldData,
+          resource: _.cloneDeep(dialogFieldData),
         },
```
Comment (AparnaKarve) [inline client/app/states/marketplace/details/details.state.js]: @himdel Would it be possible to inject `lodash` and call `cloneDeep` via `lodash`?
I believe that is what is done in other places, so just to keep things uniform...

File: client/app/services/shopping-cart.service.js [hunk 3]
```diff
@@ -31,6 +32,14 @@
       notify();
     }
 
+    function removeItem(item) {
+      state.items = _.filter(state.items, function(i) {
```
Comment (AparnaKarve) [inline client/app/services/shopping-cart.service.js]: Also here. Can you call `.filter` via `lodash` ?

File: client/app/services/shopping-cart.service.js [hunk 4]
```diff
@@ -19,6 +19,12 @@
     };
 
     var persistence = {
+      theCard: function() {
+        // api/service_orders?filter[]=state=cart
```
Comment (AparnaKarve) [inline client/app/services/shopping-cart.service.js]: probably a typo... `theCard` or `theCart`?

**Comment by AparnaKarve** [comment]
Overall, the UI looks great - but you do need to take RBAC into account.
For e.g. -
 (a) do not display the shopping cart icon 
(b) do not display `Add to shopping cart` button 
if the user is not entitled for that feature

**Comment by AparnaKarve** [comment]
@dclarizio @himdel I will resume testing once the APIs are in.

**Comment by himdel** [comment]
Thanks @AparnaKarve, I'll fix the `lodash` thing, good catch! (Though, I *may* do a `s/lodash/_/` PR later, because that's how we're using it in the main UI.)

As for RBAC.. do we actually have a feature for that? Or should I add one? Or maybe use the same feature that shows/hides the section with the "add to cart" buttons? (That would be `svc_catalog_provision`, I think.)

**Comment by AparnaKarve** [comment]
@himdel But don't you think `lodash.someThing()` reads better than `_.someThing()` ?

Yes, I think we should use the same feature id that lets us order services, which would be `svc_catalog_provision`

File: client/app/services/shopping-cart.service.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    function reload() {
+      CollectionsApi.query('service_requests', {
+        expand: 'resources',
+        filter: [ 'process=false' ],
+      })
+      .then(function(response) {
+        console.log('reload', response);
+
```
Comment (AparnaKarve) [inline client/app/services/shopping-cart.service.js]: Maybe a leftover `console.log` here.

**Comment by dclarizio** [comment]
@himdel doesn't `Add to Cart` replace `Submit Request` as the new way to order a Service?  I'm thinking we don't need the submit button anymore.

**Comment by himdel** [comment]
@dclarizio Could be, I'll remove it :)

**Comment by himdel** [comment]
Should be ready now, removing WIP...

@AparnaKarve could you re-review please? (You will need ManageIQ/manageiq#7971.)

**Comment by AparnaKarve** [comment]
@himdel This works great for the most part. One of the things that I noticed could be API related.
Here's what I found so far -

- Some Service Catalogs could not be ordered and instead gave this error message -
screen shot 2016-05-06 at 4 24 52 pm

If it's a valid error message, then the Service Catalog is not orderable and perhaps should not be even shown in the list. (just thinking out loud here)

- When I order a service catalog, I see a 404 in the console
```
Ordering error
angular.js:11630 GET http://localhost:3001/api/service_orders/cart/service_requests?expand=resources 404 (Not Found)
```

- (Minor) The console.log messages like `Item added to shopping cart` should be removed

**Comment by himdel** [comment]
@aparnakarve what's generating the first error is the automate validation code.. So it should be orderable, just not with these exact values..  But yeah I'd rather we could run some validation during, not after, but.. Api limitation :-)
And the same goes for that 404, when you just ordered, the shipping cart doesn't exist anymore and needs to be created again, hence the 404.

As for the console output, I'll check, but I think Notification is doing that at the same time it displays it to the user...

**Pull request merged**